### PR TITLE
add wait.PollImmediate() and retool wait tests

### PR DESCRIPTION
**Pull request by thockin**
Broke this out from the e2e test PR #14957

### CODE REVIEW

**Comment by thockin** [comment]
@ncdc @smarterclayton

File: pkg/util/wait/wait.go [hunk 1]
```diff
@@ -59,16 +75,16 @@ func PollInfinite(interval time.Duration, condition ConditionFunc) error {
...
 type WaitFunc func() <-chan struct{}
 
-// WaitFor gets a channel from wait(), and then invokes c once for every value
-// placed on the channel and once more when the channel is closed.  If c
-// returns an error the loop ends and that error is returned, and if c returns
+// WaitFor gets a channel from wait(), and then invokes fn once for every value
+// placed on the channel and once more when the channel is closed.  If fn
+// returns an error the loop ends and that error is returned, and if fn returns
 // true the loop ends and nil is returned. ErrWaitTimeout will be returned if
-// the channel is closed without c ever returning true.
```
Comment (bprashanth) [inline pkg/util/wait/wait.go]: > ErrWaitTimeout will be returned if the channel is closed without c ever returning true.

Drive by nit, this is one of my favorite errors: timed out waiting for the condition.
Can we please `return fmt.Errorf("%v: %v", ErrWaitTimeout, err)`, though that might break anyone checking equality with ErrWaitTimeout.

**Comment by ncdc** [comment]
With go 1.4.2 on Fedora 21:

```
$ go test -c -race ./pkg/util/wait
vagrant@localhost:~/go/src/k8s.io/kubernetes (thockin-wait-poll) ./wait.test
==================
WARNING: DATA RACE
Read by goroutine 14:
  _/home/vagrant/go/src/k8s.io/kubernetes/pkg/util/wait.TestPoll()
      /home/vagrant/go/src/k8s.io/kubernetes/pkg/util/wait/wait_test.go:86 +0x392
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:447 +0x133

Previous write by goroutine 15:
  _/home/vagrant/go/src/k8s.io/kubernetes/pkg/util/wait.func·003()
      /home/vagrant/go/src/k8s.io/kubernetes/pkg/util/wait/wait_test.go:55 +0x10c

Goroutine 14 (running) created at:
  testing.RunTests()
      /usr/lib/golang/src/testing/testing.go:555 +0xd4e
  testing.(*M).Run()
      /usr/lib/golang/src/testing/testing.go:485 +0xe0
  main.main()
      _/home/vagrant/go/src/k8s.io/kubernetes/pkg/util/wait/_test/_testmain.go:60 +0x28c

Goroutine 15 (finished) created at:
  _/home/vagrant/go/src/k8s.io/kubernetes/pkg/util/wait.func·004()
      /home/vagrant/go/src/k8s.io/kubernetes/pkg/util/wait/wait_test.go:59 +0x184
  _/home/vagrant/go/src/k8s.io/kubernetes/pkg/util/wait.WaitFor()
      /home/vagrant/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:84 +0x43
  _/home/vagrant/go/src/k8s.io/kubernetes/pkg/util/wait.pollInternal()
      /home/vagrant/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:52 +0x51
  _/home/vagrant/go/src/k8s.io/kubernetes/pkg/util/wait.TestPoll()
      /home/vagrant/go/src/k8s.io/kubernetes/pkg/util/wait/wait_test.go:80 +0x123
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:447 +0x133
==================
```

**Comment by ncdc** [comment]
Granted, I haven't gotten the race failures using `KUBE_RACE=-race ./hack/test-go.sh pkg/util/wait`

**Comment by ncdc** [comment]
Other than a potential race in the test code, and the comment from @bprashanth, provisional LGTM :smile:

**Comment by thockin** [comment]
Yeah, I saw teh race when writing it and forgot to fix.  Now with atomics.

File: pkg/util/wait/wait.go (hunk 1 above)
Comment (thockin) [inline pkg/util/wait/wait.go]: we can do that as a different PR.  I'll file a bug.

**Comment by thockin** [comment]
@k8s-oncall you might manually merge this - it should make tests run significantly faster, to help keep merge-bot moving

**Pull request merged**